src/vulkan/pipeline.cc

```diff
@@ -228,6 +273,58 @@ Result Pipeline::UpdateDescriptorSetsIfNeeded() {
   return {};
 }
 
+void Pipeline::PushConstants() {
+  if (push_constant_data_.empty())
+    return;
+
+  auto push_const_range = GetPushConstantRange();
+
+  HandleValueWithMemory memory_updater;
+  std::vector<uint8_t> memory(push_const_range.offset + push_const_range.size);
+  for (const auto& data : push_constant_data_) {
+    memory_updater.UpdateMemoryWithData(memory.data(), data);
+  }
+
+  // Based on spec, offset and size in bytes of push constant must
+  // be multiple of 4.
+  assert(push_const_range.offset % 4U == 0 && push_const_range.size % 4U == 0);
+
+  vkCmdPushConstants(command_->GetCommandBuffer(), pipeline_layout_,
+                     VK_SHADER_STAGE_ALL, push_const_range.offset,
+                     push_const_range.size, &memory[push_const_range.offset]);
+}
+
+Result Pipeline::AddPushConstant(const BufferCommand* command) {
+  if (!command->IsPushConstant())
+    return Result(
+        "Pipeline::AddPushConstant BufferCommand type is not push constant");
+
+  // TODO(jaebaek): push constant -> compute -> push constant -> compute
+  //                The second compute maybe have different push constant
+  //                ranges. Doublecheck if we update |pipeline_layout_| properly
+  //                in that case.
+  if (descriptor_related_objects_already_created_) {
+    for (auto& info : descriptor_set_info_) {
+      vkDestroyDescriptorSetLayout(device_, info.layout, nullptr);
+      if (info.empty)
+        continue;
+
+      vkDestroyDescriptorPool(device_, info.pool, nullptr);
+    }
+
+    vkDestroyPipelineLayout(device_, pipeline_layout_, nullptr);
+    vkDestroyPipeline(device_, pipeline_, nullptr);
+
+    descriptor_related_objects_already_created_ = false;
+  }
+
+  push_constant_data_.push_back({command->GetDatumType().GetType(),
```

Comment:
Suggest emplace_back here since you're doing the initialization and a copy, and emplace_back would save you the copy.

Reply:
done